relay-server/src/actors/upstream.rs

```diff
@@ -235,1043 +252,1229 @@ impl fmt::Display for RequestPriority {
-            } else if let Some(msg) = self.low_prio_requests.pop_back() {
-                self.send_request(msg, ctx);
-            } else {
-                break; // no more messages to send at this time stop looping
-            }
-        }
+    /// Returns the number of entries in the queue.
+    pub fn len(&self) -> usize {
+        self.high.len() + self.low.len()
     }
-}
 
-struct ScheduleConnectionCheck;
+    /// Puts an entry into the queue at the given position.
+    fn put(&mut self, entry: Entry, position: EnqueuePosition) {
+        let priority = entry.request.priority();
 
-impl Message for ScheduleConnectionCheck {
-    type Result = ();
-}
+        // We can ignore send errors here. Once the channel closes, we drop all incoming requests
+        // here. The receiving end of the request will be notified of the drop if they are waiting
+        // for it.
+        let queue = match priority {
+            RequestPriority::High => &mut self.high,
+            RequestPriority::Low => &mut self.low,
+        };
 
-impl Handler<ScheduleConnectionCheck> for UpstreamRelay {
-    type Result = ();
+        match position {
+            EnqueuePosition::Front => queue.push_front(entry),
+            EnqueuePosition::Back => queue.push_back(entry),
+        }
 
-    fn handle(&mut self, _msg: ScheduleConnectionCheck, ctx: &mut Self::Context) -> Self::Result {
-        self.schedule_connection_check(ctx);
+        relay_statsd::metric!(
+            histogram(RelayHistograms::UpstreamMessageQueueSize) = self.len() as u64,
+            priority = priority.name(),
+        );
     }
-}
-
-/// Checks the status of the network connection with the upstream server
-struct GetHealthCheck;
 
-impl UpstreamRequest for GetHealthCheck {
-    fn method(&self) -> Method {
-        Method::GET
+    /// Places an entry at the back of the queue.
+    ///
+    /// Since entries are dequeued in FIFO order, this entry will be dequeued last within its
+    /// priority class.
+    pub fn enqueue(&mut self, entry: Entry) {
+        self.put(entry, EnqueuePosition::Back)
     }
 
-    fn path(&self) -> Cow<'_, str> {
-        Cow::Borrowed("/api/0/relays/live/")
+    /// Place an entry at the front of the queue.
+    ///
+    /// This entry will be dequeued next within its priority class, unless another call to
+    /// `enqueue_immediate` follows.
+    pub fn enqueue_immediate(&mut self, entry: Entry) {
+        self.put(entry, EnqueuePosition::Front)
     }
 
-    fn retry(&self) -> bool {
-        false
+    /// Removes the head of the queue with highest priority.
+    ///
+    /// This always returns entries with [high](RequestPriority::High) first before dequeueing
+    /// entries with low priority.
+    pub fn dequeue(&mut self) -> Option<Entry> {
+        self.high.pop_front().or_else(|| self.low.pop_front())
     }
+}
 
-    fn priority(&self) -> RequestPriority {
-        RequestPriority::Immediate
-    }
+/// The state of authentication state.
```

Comment:
I think it's already give the understanding what it is: 

```suggestion
/// The current authentication state.
```
Or the part `state of ... state` should be rewritten in other way, since it's a bit confusing.